postgis/lwgeom_remove_small_parts.c

```diff
@@ -0,0 +1,244 @@
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * PostGIS is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PostGIS.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **********************************************************************
+ *
+ * Copyright (C) 2024 Sam Peters <gluser1357@gmx.de>
+ *
+ **********************************************************************/
+
+#include "postgres.h"
+#include "funcapi.h"
+#include "utils/array.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+#include "utils/numeric.h"
+#include "access/htup_details.h"
+
+#include "liblwgeom.h"
+#include "liblwgeom_internal.h"
+
+// ===============================================================================
+// helper function for polygon and polyline POINTARRAY's
+// which removes small parts given by dx and dy.
+// ===============================================================================
+static void ptarray_remove_dim_helper(POINTARRAY *points, double mindx, double mindy) {
+
+    int r;
+    double xmin = 0, ymin = 0, xmax = 0, ymax = 0;
+    double x, y;
+    POINT4D point;
+
+    int npoints = points->npoints;
+    for (r=0; r < npoints; r++) {
+
+		getPoint4d_p(points, r, &point); // point read/write see ptarray_flip_coordinates
+
+		x = point.x;
+		y = point.y;
+
+		if (mindx > 0) {
+		    if (!r || xmin > x) xmin = x;
+		    if (!r || xmax < x) xmax = x;
+		}
+		if (mindy > 0) {
+		    if (!r || ymin > y) ymin = y;
+		    if (!r || ymax < y) ymax = y;
+		}
+	}
+
+    if ((mindx > 0 && xmax - xmin < mindx) ||
+    	(mindy > 0 && ymax - ymin < mindy)) {
+	    // skip part
+	    points->npoints = 0;
+    }
+}
+
+// ===============================================================================
+// remove small (sub-)geometries being smaller than given dimensions.
+// 2D-(MULTI)POLYGONs and (MULTI)LINESTRINGs are evaluated, others keep untouched.
+// ===============================================================================
+PG_FUNCTION_INFO_V1(ST_RemoveSmallParts);
+Datum ST_RemoveSmallParts(PG_FUNCTION_ARGS) {
+
+    double mindx = 0, mindy = 0;
+    unsigned int i, j, iw, jw;
+
+    // gserialized logic see for example in /postgis/lwgeom_functions_basic.c,
+    // type definitions see /liblwgeom/liblwgeom.h(.in)
+
+    GSERIALIZED *serialized_in;
+    GSERIALIZED *serialized_out;
+
+    LWGEOM *geom;
+
+    // geom input check
+    if (PG_GETARG_POINTER(0) == NULL) {
+        PG_RETURN_NULL();
+    }
+
+    serialized_in = (GSERIALIZED *)PG_DETOAST_DATUM_COPY(PG_GETARG_DATUM(0));
+
+    // dimension check (only 2d geometry types are supported yet)
+    if (gserialized_ndims(serialized_in) < 2) {
```

Comment:
Isn't this check reversed?

```suggestion
    if (gserialized_ndims(serialized_in) > 2) {
```